src/main.cpp

```diff
@@ -403,39 +403,280 @@ bool FullSyncWithDPORNodes()
+    int64_t nCurrentTime = GetAdjustedTime();
+    if (fDebug10) LogPrintf("ETTS debug: nCurrentTime = %i", nCurrentTime);
+    
+    CTxDB txdb("r");
+    // Here I am defining a time mask 16 times as long as the normal stake time mask. This is to quantize the UTXO's into a maximum of
+    // 16 hours * 3600 / 256 = 225 time bins for evaluation. Otherwise for a large number of UTXO's, this algorithm could become
+    // really expensive.
+    const int ETTS_TIMESTAMP_MASK = 16 * STAKE_TIMESTAMP_MASK;
 
-    return result/100;
+    int64_t BalanceAvailForStaking = 0;
+    vector<COutput> vCoins;
+    
+    {
+        LOCK2(cs_main, pwalletMain->cs_wallet);
+        
+        BalanceAvailForStaking = pwalletMain->GetBalance() - nReserveBalance;
+
+        if (fDebug10) LogPrintf("ETTS debug: BalanceAvailForStaking = %u", BalanceAvailForStaking);
+        
+        // Get out early if no balance available and set return value of 0. This should already have happened above, because with no
+        // balance left after reserve, staking should be disabled; however, just to be safe...
+        if (BalanceAvailForStaking <= 0)
+        {
+            if (fDebug10) LogPrintf("ETTS debug: No balance available: ETTS = %f", result);
+            return result;
+        }
+
+        //reminder... void AvailableCoins(std::vector<COutput>& vCoins, bool fOnlyConfirmed=true, const CCoinControl *coinControl=NULL, bool fIncludeStakingCoins=false) const;
+        pwalletMain->AvailableCoins(vCoins, true, NULL, true);
+    }
+
+    
+    // An efficient local structure to store the UTXO's with the bare minimum info we need.
+    typedef vector< std::pair<int64_t, int64_t> > vCoinsExt;
+    vCoinsExt vUTXO;
+    // A local ordered set to store the unique "bins" corresponding to the UTXO transaction times. We are going to use this
+    // for the outer loop.
+    std::set<int64_t> UniqueUTXOTimes;
+        // We want the first "event" to be the CurrentTime. This does not have to be quantized.
+    UniqueUTXOTimes.insert(nCurrentTime);
+
+    
+    // Debug output cooldown...
+    if (fDebug10) LogPrintf("ETTS debug: nStakeMinAge = %i", nStakeMinAge);
+
+    // Derive a smoothed difficulty over desired PoSInterval of blocks by calling GetPoSKernelPS() which is netweight and reverse engineering...
+    double dDiff = GetAverageDifficulty(nPoSInterval);
+    if (fDebug10) LogPrintf("ETTS debug: dDiff = %f", dDiff);
+    
+    // The stake probability per "throw" of 1 weight unit = target value at diff of 1.0 / (maxhash * diff). This happens effectively every STAKE_TIMESTAMP_MASK+1 sec.
+    double dUnitStakeProbability = 1 / (4295032833.0 * dDiff);
+    if (fDebug10) LogPrintf("ETTS debug: dUnitStakeProbability = %e", dUnitStakeProbability);
+
+    
+    int64_t nTime = 0;
+    for (const auto& out : vCoins)
+    {
+        CTxIndex txindex;
+        CBlock CoinBlock; //Block which contains CoinTx
+        {
+            LOCK2(cs_main, pwalletMain->cs_wallet);
```

Comment:
I don't think we need to lock here.

Reply:
I followed what was in miner.cpp on that one. I will try without the lock and see what happens.

Reply:
Looks like it works without the lock.